### PR TITLE
bazel: update to 4.1.0

### PR DESCRIPTION
**Pull request by junr03**
Description: updates to bazel 4.1.0 by removing deprecated dependencies and bespoke build rules for swift.
Risk Level: low
Testing: CI

Signed-off-by: Jose Nino <jnino@lyft.com>

### CODE REVIEW

**Comment by junr03** [comment]
Heavy _inspiration_ from  https://github.com/envoyproxy/envoy-mobile/pull/797

File: bazel/envoy_mobile_repositories.bzl [hunk 1]
```diff
@@ -15,24 +15,6 @@ def envoy_mobile_repositories():
     python_repos()
 
 def upstream_envoy_overrides():
```
Comment (goaway) [inline bazel/envoy_mobile_repositories.bzl:17]: Can this function be removed altogether if it's empty?

File: library/swift/BUILD [hunk 2]
```diff
@@ -1,9 +1,11 @@
...
+load("@build_bazel_rules_apple//apple:ios.bzl", "ios_static_framework")
+load("@build_bazel_rules_swift//swift:swift.bzl", "swift_library")
+load("//bazel:swift_header_collector.bzl", "swift_header_collector")
 
 licenses(["notice"])  # Apache 2
 
-swift_static_framework(
-    name = "ios_framework",
+swift_library(
+    name = "lib",
```
Comment (goaway) [inline library/swift/BUILD:8]: Any particular reason for this name (e.g. like does it align with some convention or is it required by some rule)? If not, as a nit, we might want to pick something a little more explicit.

**Comment by goaway** [comment]
Nice! Thank you @junr03!

File: bazel/swift_header_collector.bzl [hunk 3]
```diff
@@ -0,0 +1,21 @@
+"""
+Propagate the generated Swift header from a swift_library target
+This exists to work around https://github.com/bazelbuild/rules_swift/issues/291
+"""
+
+def _swift_header_collector(ctx):
+    return [
+        DefaultInfo(
+            files = ctx.attr.library[CcInfo].compilation_context.headers,
```
Comment (keith) [inline bazel/swift_header_collector.bzl:9]: might want to assert that this only has 1 file to make sure it is what we expect, not a huge deal

File: bazel/swift_static_framework.bzl [hunk 4]
```diff
@@ -1,218 +0,0 @@
-"""
```
Comment (keith) [inline bazel/swift_static_framework.bzl:1]: awesome!

File: examples/objective-c/hello_world/ViewController.m [hunk 5]
```diff
@@ -1,4 +1,4 @@
-#import <Envoy/Envoy-Swift.h>
```
Comment (keith) [inline examples/objective-c/hello_world/ViewController.m:1]: not sure if you consider this public API or not

**Review by keith — DISMISSED** [review]

File: bazel/envoy_mobile_repositories.bzl (hunk 1 above)
Comment (junr03) [reply bazel/envoy_mobile_repositories.bzl:17]: not empty. There is more stuff below

File: library/swift/BUILD [hunk 6]
```diff
@@ -44,3 +46,17 @@ swift_static_framework(
     private_deps = ["//library/objective-c:envoy_engine_objc_lib"],
     visibility = ["//visibility:public"],
```
Comment (keith) [inline library/swift/BUILD:47]:
```suggestion
    visibility = ["//visibility:public"],
    features = ["swift.enable_library_evolution"],
```

**Comment by junr03** [comment]
Awesome! Green. Can I get your eyes @goaway @keith ?

**Review by keith — APPROVED** [review]

File: bazel/envoy_mobile_repositories.bzl (hunk 1 above)
Comment (goaway) [reply bazel/envoy_mobile_repositories.bzl:17]: ah, I see, the GitHub UI threw me.

**Review by goaway — APPROVED** [review]
Thanks @junr03 and @keith!

**Pull request merged**